File: transpiler/javatests/com/google/j2cl/readable/java/arithmeticexpressions/ArithmeticExpressions.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package arithmeticexpressions;

import org.jspecify.annotations.Nullable;

public class ArithmeticExpressions {
  private static final float FLOAT_CONSTANT = 1.1f;
  private static final double DOUBLE_CONSTANT = FLOAT_CONSTANT;
  private static final double DOUBLE_CONSTANT_WITH_ARITHMETIC = FLOAT_CONSTANT + FLOAT_CONSTANT;

  public void testCoercions() {
    byte b = (byte) 1L;
    char c = (char) 1L;
    short s = (short) 1L;
    int i = (int) 1L;
    float f = (float) 1L;
    double d = (double) 1L;
    b = (byte) 9223372036854775807L; // -1
    c = (char) 9223372036854775807L; // 65535
    s = (short) 9223372036854775807L; // -1
    i = (int) 9223372036854775807L; // -1
    f = (float) 9223372036854775807L; //  9.223372036854776E18
    d = (double) 9223372036854775807L; //  9.223372036854776E18
    Object o = c;
    s = (short) (char) o;
  }

  public void testPrimitives() {
    int a = 10,
        b = a++,
        c = a--,
        d = ++a,
        e = --a,
        f = -a,
        g = +a,
        h = ~a,
        i = 1 + 1 + 2 - 5,
        j = (1 + 2) * (3 + 4),
        p = 1 / 2 * 3 % 4,
        r = -0x80000000,
        t = -(-(-1)),
        u = +(+(+1)),
        v = -(+(-1)),
        w = 5 - (-4);
    boolean k = !(1 + 2 + 3 == 4), l = (1 + 2 != 4), m = Long.MAX_VALUE != 9223372036854776833d;
    double o = ((5 + 1) / 2) - 0.0;

    a = a << 31L;
    a <<= 1L;
    a += 1L;
    a /= 1L;
    a += Double.MAX_VALUE; // if not expanded and performed in double result will change.

    ((k)) |= true;

    short s = 10;
    k = ++s == 10;

    int q = 3 >> 2;
    q = 3 >>> 2;

    byte x = (byte) (((short) a + (byte) (short) b) * (short) (byte) c);

    char y = 'y';
    Long z = 121L;
    k = y == z;
  }

  // Compount assignments in static fields
  public static long one = 1;
  public static long foo = one++;
  public long bar = foo++;

  public void testCompoundArray() {
    int[] ints = null;
    ints[0] += 1;
    ints[0] -= 1;
    ints[0] *= 1;
    ints[0] /= 1;
    ints[0] &= 1;
    ints[0] ^= 1;
    ints[0] |= 1;
    ints[0] %= 1;
    ints[0] <<= 1;
    ints[0] >>= 1;
    ints[0] >>>= 1;
    ints[0]++;
    ++ints[0];
    int i = 0;
    ints[i++]++;
    ++ints[++i];
    ints[i++] /= 1;

    long[] longs = null;
    longs[0] += 1;
    longs[0]--;
    --longs[0];
    getLongArray()[0]++;

    boolean[] booleans = null;
    booleans[0] |= true;

    String[] strings = null;
    strings[0] += null;

    short[] shorts = null;
    boolean b = ++shorts[0] == 10;
  }

  private static long[] getLongArray() {
    return null;
  }

  public void testCompoundBoxedTypes() {
    Integer c = 1000;

    // Compound expressions.
    Integer d = 10000;
    d += c;

    int i = 43;
    d += i;
    d <<= i;
    i += c;

    // Prefix expressions;
    Integer e = ++c;
    e = ++c;
    Double e2 = 100d;
    ++e2;

    // Postfix expressions.
    Integer f = c++;
    f = c++;
    Byte b = 0;
    b++;
    Character ch = 'c';
    ch++;

    // Method call on result of arithmetic expressions.
    (++f).intValue();
    (f--).intValue();

    class Ref<T> {
      @Nullable T field;
    }

    Ref<Byte> ref = null;
    ref.field++;
    int n = 1 + ref.field;
  }

  private static Integer getInteger() {
    return null;
  }

  long intField;

  private static void testSideEffect() {
    getWithSideEffect().intField += 5;
  }

  private static ArithmeticExpressions getWithSideEffect() {
    return null;
  }

  // This is a readable example that exposes the nuances of JDT representation of InfixExpression,
  // where certain expressions are represented as operator, list of operands rather than a
  // nested binary tree structure.
  public void testExtendedOperands() {
    Integer boxedInteger = 3;
    int i;
    long l;
    double d;
    l = 2 - boxedInteger - 2L;
    l = 2 | boxedInteger | 2L;
    l = 1000000L * l * 60 * 60 * 24;
    l = 24 * 60 * 60 * l * 1000000L;
    d = l = i = 20;
    l = boxedInteger = i = 20;
    l = i + boxedInteger + l + 20;
    d = 20 + l + d;
  }

  public void testEffectivelyFinalVariableInAssignmentExpression(boolean condition) {
    long effectivelyFinal;
    // "condition &&" is necessary to trigger a problem from b/251115461
    if (condition && ((effectivelyFinal = bar) != 0)) {
      long unused = effectivelyFinal;
    }
  }

  private static long counter = 0;

  private static long incrementCounter() {
    return ++counter;
  }
}

File: transpiler/javatests/com/google/j2cl/readable/kotlin/arithmeticexpressions/ArithmeticExpressions.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific languagCS-ArithmeticExpressions-2022-08-12_115430e governing permissions and
 * limitations under the License.
 */
package arithmeticexpressions

private const val FLOAT_CONSTANT = 1.1f
private const val DOUBLE_CONSTANT = FLOAT_CONSTANT
private const val DOUBLE_CONSTANT_WITH_ARITHMETIC = FLOAT_CONSTANT + FLOAT_CONSTANT

fun testCoercions() {
  var b: Byte = 1L.toByte()
  var c: Char = 1L.toChar()
  var s: Short = 1L.toShort()
  var i: Int = 1L.toInt()
  var f: Float = 1L.toFloat()
  var d: Double = 1L.toDouble()
  b = 9223372036854775807L.toByte() // -1
  c = 9223372036854775807L.toChar() // 65535
  s = 9223372036854775807L.toShort() // -1
  i = 9223372036854775807L.toInt() // -1
  f = 9223372036854775807L.toFloat() //  9.223372036854776E18
  d = 9223372036854775807L.toDouble() //  9.223372036854776E18
  val o: Any = c
  s = (o as Char).toShort()
}

fun testPrimitives() {
  var a = 10
  val b = a++
  val c = a--
  val d = ++a
  val e = --a
  val f = -a
  val g = +a
  val h = a.inv()
  val i = 1 + 1 + 2 - 5
  val j = (1 + 2) * (3 + 4)
  val p = 1 / 2 * 3 % 4
  val r = -0x80000000
  val t = -(-(-1))
  val u = +(+(+1))
  val v = -(+(-1))
  val w = 5 - (-4)
  var k = !(1 + 2 + 3 == 4)
  val l = (1 + 2 != 4)
  val m = Long.MAX_VALUE != 9223372036854776833.0.toLong()
  val o: Double = ((5 + 1) / 2) - 0.0

  a = a shr 31
  a = a shr 1
  a += 1
  a /= 1
  a += Double.MAX_VALUE.toInt() // if not expanded and performed in double result will change.

  k = k or true

  var s: Short = 10
  k = ++s == 10.toShort()

  var q = 3 shr 2
  q = 3 ushr 2

  val x: Byte = ((a.toShort() + b.toShort().toByte()) * c.toByte().toShort()).toByte()

  val y = 'y'
  val z = 121L
  k = y.toLong() == z
}

fun testDoubleAssignments() {
  var a = 1.0
  a += 2.0
  a -= 3.0
  a *= 4.0
  a /= 5.0
  a %= 6.0
}

// Compount assignments in static fields
var one: Long = 1L
var foo = one++
val bar = foo++

fun testCompoundArray() {
  val ints = IntArray(0)
  ints[0] += 1
  ints[0] -= 1
  ints[0] *= 1
  ints[0] /= 1
  ints[0] %= 1
  ints[0]++
  ++ints[0]
  var i = 0
  ints[i++]++
  ++ints[++i]
  ints[i++] /= 1

  val longs = LongArray(0)
  // Missing the corresponding set method for array of longs. Trying to compile
  //     longs[0] += 1
  // gives "no set method providing array access"
  longs[0] += 1L
  longs[0]--
  --longs[0]
  getLongArray()[0]++

  val strings = null as Array<String?>?
  // strings[0] += null

  val shorts = ShortArray(0)
  val b = ++shorts[0] == 10.toShort()
}

fun testCompoundBoxedTypes() {
  class Ref<T>(var field: T)

  var ref = Ref<Byte>(1)
  ref.field++
  ++ref.field
  val n = 1 + ref.field

  var longRef = Ref(1L)
  longRef.field += 1
}

fun getLongArray(): LongArray {
  return LongArray(0)
}

class A {
  var intField: Int = 1
}

private fun getInteger(): Int? {
  return null
}

fun testSideEffect() {
  getWithSideEffect().intField += 5
}

fun getWithSideEffect(): A {
  return A()
}

var counter = 0

fun incrementCounter(): Int {
  return ++counter
}

fun testEquals() {
  val nullableDouble: Double? = 1.0
  val double = 1.0
  val nullableFloat: Float? = 1.0f
  val float = 1.0f

  // These are ieee754 semantics
  var b = nullableDouble == double
  b = double == nullableDouble
  b = double == double
  b = nullableDouble == nullableDouble

  b = nullableFloat == float
  b = float == nullableFloat
  b = float == float
  b = nullableFloat == nullableFloat

  // These are regular equals semantics
  b = double == null
  b = nullableDouble == null
  b = null == double
  b = null == nullableDouble
  b = float == null
  b = nullableFloat == null
  b = null == float
  b = null == nullableFloat

  val o1: Any? = Any()
  val o2: Any? = "Hello"
  b = o1 == o2
}

fun testTripleEquals() {
  val nullableDouble: Double? = 1.0
  val double = 1.0
  val nullableFloat: Float? = 1.0f
  val float = 1.0f

  var b = nullableDouble === double
  b = double === nullableDouble
  b = double === double
  b = nullableDouble === nullableDouble

  b = nullableFloat === float
  b = float === nullableFloat
  b = float === float
  b = nullableFloat === nullableFloat

  b = double === null
  b = nullableDouble === null
  b = null === double
  b = null === nullableDouble
  b = float === null
  b = nullableFloat === null
  b = null === float
  b = null === nullableFloat

  val o1: Any? = Any()
  val o2: Any? = "Hello"
  b = o1 === o2
}
